### DISCUSSION TITLE
New RFID RC522 Template Creation

### DISCUSSION THREAD

**Discussion by wlan4you** [arendst/Tasmota#10330] [Q&A] [answered]
Edit: Edit PinOut Picture from RC522

Hello together,
i try to bring the new RC522 RFID running with tasmota.
May be somebody have some TIPS for a **BEGINNER** like me with Node-Red, MQTT and TASMOTA
I use a **ESP8266 NODEMCU AMICA** Pinout see attached.

I will use it with a: 
1. Temp & Humi AM2301 ✅
2. Scale HX711 ✅
3. RFID RC522 ❌

I flashed the new Version from today **V9.2.0.2**.
What Pins from the RC522 must i connect to the ESP8266,
and what must i configure in the Template parameters to see the RFID?
I can not find the new Change from ``GPIO_SPI_CS`` by ``GPIO_RC522_CS`` in the templates.

Is that enough to activate the:
  #define USE_RC522                              // Add support for MFRC522 13.56Mhz Rfid reader (+6k code)
    #define USE_RC522_DATA_FUNCTION              // Add support for reading data block content (+0k4 code)
    #define USE_RC522_TYPE_INFORMATION           // Add support for showing card type (+0k4 code)
in the my_user_config.h
And build the binary?

Thank's for help me.
All the best for 2021!!!

In the new Changelog are:
## [9.2.0.2]
.....
### Breaking Changed
- Replaced MFRC522 13.56MHz rfid card reader GPIO selection from ``GPIO_SPI_CS`` by ``GPIO_RC522_CS``
- Replaced ILI9341 GPIO selection from ``GPIO_SPI_CS`` by ``GPIO_ILI9341_CS`` and ``GPIO_SPI_DC`` by ``GPIO_ILI9341_DC``

![Tasmota1](https://user-images.githubusercontent.com/60288714/103408331-49aec200-4b62-11eb-9f92-02bec15d3a32.JPG)
![Tasmota2](https://user-images.githubusercontent.com/60288714/103408336-516e6680-4b62-11eb-8f5f-740f2497531f.JPG)
![Tasmota5](https://user-images.githubusercontent.com/60288714/103408357-7bc02400-4b62-11eb-8ab0-8d8bb7455ebf.JPG)
![nodemcu_pins](https://user-images.githubusercontent.com/60288714/103408368-824e9b80-4b62-11eb-952e-60431a9dab02.png)
![Tasmota6](https://user-images.githubusercontent.com/60288714/103408527-2cc6be80-4b63-11eb-9a79-2b9d72867c5d.JPG)
![RC522_pins](https://user-images.githubusercontent.com/60288714/103409251-343b9700-4b66-11eb-81e4-e249acafc23e.png)

**Comment by arendst** [comment, marked as answer]
Do NOT use SSPI GPIOs as the MFRC522 driver only supports hardware SPI.

You only need to select GPIO_RC522_CS prefferably on GPIO15. Tasmota then will select GPIO12-14 as hardware SPI pins.

**Reply by wlan4you** [reply]
> GPIO_RC522_CS

GPIO_RC522_CS
But it not comes on the Pulldownlist @ Template.
Here i have more Information what i have done.
https://github.com/arendst/Tasmota/discussions/10330#discussioncomment-254315
I do it again, but it isn't in the List.

**Comment by barbudor** [comment]
Hi @wlan4you 
As explained on Discord, you need to compile your own custom firmware.
Please explain what you have done.

**Reply by wlan4you** [reply]
What i have done:
1. Download the complete new ZIP from Today with Tasmota-development
2. Edit: my_user_config.h (Some Settings to WIFI, MQTT, Name of Device,...)
**AND ROW 631, 632 and 633 activated** like:
// -- SPI sensors ---------------------------------
//#define USE_SPI                                  // Hardware SPI using GPIO12(MISO), GPIO13(MOSI) and GPIO14(CLK) in addition to two user selectable GPIOs(CS and DC)

#ifdef USE_SPI
//  #define USE_NRF24                              // Add SPI support for NRF24L01(+) (+2k6 code)
    #define USE_MIBLE                            // BLE-bridge for some Mijia-BLE-sensors (+4k7 code)
//  #define USE_DISPLAY                            // Add SPI Display support for 320x240 and 480x320 TFT
    #define USE_DISPLAY_ILI9341                  // [DisplayModel 4] Enable ILI9341 Tft 480x320 display (+19k code)
//    #define USE_DISPLAY_EPAPER_29                // [DisplayModel 5] Enable e-paper 2.9 inch display (+19k code)
//    #define USE_DISPLAY_EPAPER_42                // [DisplayModel 6] Enable e-paper 4.2 inch display
//    #define USE_DISPLAY_ILI9488                  // [DisplayModel 8] [I2cDriver38] (Touch)
//    #define USE_DISPLAY_SSD1351                  // [DisplayModel 9]
//    #define USE_DISPLAY_RA8876                   // [DisplayModel 10] [I2cDriver39] (Touch)
  #define USE_RC522                              // Add support for MFRC522 13.56Mhz Rfid reader (+6k code)
    #define USE_RC522_DATA_FUNCTION              // Add support for reading data block content (+0k4 code)
    #define USE_RC522_TYPE_INFORMATION           // Add support for showing card type (+0k4 code)

#endif  // USE_SPI

4. For my Scale i activated: **Row 755 and 756 remove the "//" on each two row**.
5. make the new binary
6. Flash it with tasmotize (Erase before flashing, Self-resetting device (NodeMCU, Wemos) "acitive"
7. after Reboot, go to the IP from NodeMCU go to "CONFIGURE Template.
I try to setup like the Screenshot. **AM2301 and my HX711 SCALE are running well.**
8. I try to connect the Following Pins to 8266
I used the:
ESP8266 ----- RC522

D8 GPIO15 ----- SDA
D5 GPIO14 ----- SCK
D7 GPIO13 ----- MOSI 
D6 GPIO12 ----- MISO
-not Used- ----- IRQ (I read anywhere, that is not nessecary)
GND ----- GND
- not used- ----- RST (I read anywhere, that is not nessecary)
3.3V ----- 3.3V

And here i think my problems:
- The right PINOUT between 8266 and RC522
--- What are doing with: IRQ and RST ???
- I didn't find the right Names for the GPIO's in templates

**Comment by barbudor** [comment]
```//#define USE_SPI // Hardware SPI using GPIO12(MISO), GPIO13(MOSI) and GPIO14(CLK) in addition to two user selectable GPIOs(CS and DC)```

Did you uncommented the `#define USE_SPI`
If you don't, then RC522 will not be added because of the `#ifde USE_SPI...`

**Comment by barbudor** [comment]
With the right options, build like a charm : 
![image](https://user-images.githubusercontent.com/2996042/103414685-40801e00-4b7f-11eb-838b-0e07a0e34418.png)

```
// -- SPI sensors ---------------------------------
#define USE_SPI                                  // Hardware SPI using GPIO12(MISO), GPIO13(MOSI) and GPIO14(CLK) in addition to two user selectable GPIOs(CS and DC)

#ifdef USE_SPI
//  #define USE_NRF24                              // Add SPI support for NRF24L01(+) (+2k6 code)
    #define USE_MIBLE                            // BLE-bridge for some Mijia-BLE-sensors (+4k7 code)
//  #define USE_DISPLAY                            // Add SPI Display support for 320x240 and 480x320 TFT
    #define USE_DISPLAY_ILI9341                  // [DisplayModel 4] Enable ILI9341 Tft 480x320 display (+19k code)
//    #define USE_DISPLAY_EPAPER_29                // [DisplayModel 5] Enable e-paper 2.9 inch display (+19k code)
//    #define USE_DISPLAY_EPAPER_42                // [DisplayModel 6] Enable e-paper 4.2 inch display
//    #define USE_DISPLAY_ILI9488                  // [DisplayModel 8] [I2cDriver38] (Touch)
//    #define USE_DISPLAY_SSD1351                  // [DisplayModel 9]
//    #define USE_DISPLAY_RA8876                   // [DisplayModel 10] [I2cDriver39] (Touch)
  #define USE_RC522                              // Add support for MFRC522 13.56Mhz Rfid reader (+6k code)
    #define USE_RC522_DATA_FUNCTION              // Add support for reading data block content (+0k4 code)
    #define USE_RC522_TYPE_INFORMATION           // Add support for showing card type (+0k4 code)

#endif  // USE_SPI
```

**Reply by barbudor** [reply]
I don't have the hardware ready to test. May be later this week.
Let me know if it works for you

**Comment by barbudor** [comment]
Note that  it is not recommended to modify `my_user_config.h` but to use `user_config_override.h` to undef, define or redefine options

**Comment by wlan4you** [comment]
Hello that was the FAULT. We talk in discord over it, but somebody read only here.
I OVERREAD the #define USE_SPI

Yes, i think i understand how the user_config_override.h work.
I can there define my parameters wifi, ssid, password and so on and i must not make the big user_config everytime new.
i copy my ... override in the file to create.
I think it is solved here.
The Rest over discord !

**Reply by newmir60** [reply]
Hi, 

not sure if I am allowed to add a question to this issue (otherwise sorry). I have also an issue with RS522. I did everything according to the description in this link: https://tasmota.github.io/docs/MFRC522/.

I have tried to connect SDA to GPIO 4,5 and 15 with the same result (I did not fully understand the remark about GPIO15). As long as I dont connect SDA I get a message in the serial monitor that RC522 is not detected (not a suprise so far). Then apart from RS522 everything works. I can use A0 and interal LED (for testing) and see it working. It connects stable to my WIFI and MQTT Brokker. Even when I connect SDA physically, but not yet in the "Configuration Menu" it works as already described. But as soon as I connect SDA correctly in the "Configuration Menu" I get a problem with my WIFI. It detects RS522 now correctly and in the serial Monitor I can see RC522 working correctly, but it now does not connect any longer to my WIFI. I thought it might be a power issue, but even with a very stable power supply I have the same problem ..... has this problem been observed before? Is someone using RC522 with Tasmota succesfully? Could it be an issue of my specific RC522 board? 

Update: I have a suspicion - I have connected RC522 to 3.3 Volt and it could be better in some case to use the 5 Volt out from ESP8266  .... I'll be back on this.  

Solved: RC522 with a conection to 5V pin runs much more stable as before. This was a HW issue and has nothing to do with Tasmota.

**Reply by barbudor** [reply]
> I  did not fully understand the remark about GPIO15

I2C pins (SCL and SDA) must have a pull-up as required by the I2C bus specifications
GPIO15 must have a pull-down to allow the ESP to proer boot. So GPIO15 is not suitable as an I2C GPIO.
